### PR TITLE
Document our crate organization in ARCHITECTURE.md

### PR DESCRIPTION
**Pull request by abey79**
### What

Add a list of crates and an org diagram in ARCHITECTURE.md. Also add a lint to minimise the chance of rot.

Source FigJam: https://www.figma.com/file/Umob8ztK1HmYKLUMSq8aPb/Crates-org

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Full build: [app.rerun.io](https://app.rerun.io/pr/4458/index.html)
  * Partial build: [app.rerun.io](https://app.rerun.io/pr/4458/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json) - Useful for quick testing when changes do not affect examples in any way
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4458)
- [Docs preview](https://rerun.io/preview/61649d5f937007affd7624aa6b75d9a72dc654c3/docs) 
- [Examples preview](https://rerun.io/preview/61649d5f937007affd7624aa6b75d9a72dc654c3/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: scripts/lint.py [hunk 1]
```diff
@@ -748,6 +749,23 @@ def lint_file(filepath: str, args: Any) -> int:
     return num_errors
 
 
+def lint_crate_docs(should_ignore: Callable[[Any], bool]) -> int:
+    """Make sure ARCHITECTURE.md talks about every single create we have."""
```
Comment (emilk) [inline scripts/lint.py:753]:
```suggestion
    """Make sure ARCHITECTURE.md talks about every single crate we have."""
```

File: scripts/lint.py [hunk 2]
```diff
@@ -748,6 +749,23 @@ def lint_file(filepath: str, args: Any) -> int:
...
+
+    crates_dir = Path(__file__).parent.parent / "crates"
```
Comment (emilk) [inline scripts/lint.py:755]: This isn't very robust, considering this very PR makes me want to split `crates` into multiple sub-directories.

I suggest you find the list of crates in some other way, e,.g. folder names containing `Cargo.toml` (ignoring the root/workspace).

File: scripts/lint.py [hunk 3]
```diff
@@ -748,6 +749,23 @@ def lint_file(filepath: str, args: Any) -> int:
...
+    architecture_md_file = Path(__file__).parent.parent / "ARCHITECTURE.md"
+
+    architecture_md = architecture_md_file.read_text()
+
+    error_count = 0
+    for crate in crates_dir.iterdir():
+        if not should_ignore(crate) and crate.name not in architecture_md:
+            print(f"Missing documentation for crate {crate.name} in ARCHITECTURE.md")
```
Comment (emilk) [inline scripts/lint.py:763]: This omits the negated test, i.e. that `ARCHITECTURE.md` doesn't mention crates that no longer exist

**Review by emilk — CHANGES_REQUESTED** [review]
I suggest moving the crate list and image to the bottom of `ARCHITECTURE.md` - it takes a lot of space now. Here it is rendered: https://github.com/rerun-io/rerun/blob/antoine/crate-diagram/ARCHITECTURE.md

File: scripts/lint.py (hunk 3 above)
Comment (abey79) [reply scripts/lint.py]: Done

File: scripts/lint.py (hunk 2 above)
Comment (abey79) [reply scripts/lint.py]: Done

**Review by emilk — APPROVED** [review]

**Pull request merged**